vault/login_mfa.go

```diff
@@ -2029,23 +2035,65 @@ func (c *Core) validateOkta(ctx context.Context, mConfig *mfa.Config, username s
 		return fmt.Errorf("no MFA factors found for user")
 	}
 
-	var factorFound bool
+	var passcode string
+	if mfaFactors != nil {
+		passcode = mfaFactors.passcode
+	}
+
+	// When the user provides a passcode, we need to try each of their token* factors
+	// because we do not know which factor generated the token.
+	var passcodeUserFactorErrors map[*okta.UserFactor]error
+
 	var userFactor *okta.UserFactor
 	for _, factor := range factors {
 		if factor.IsUserFactorInstance() {
 			userFactor = factor.(*okta.UserFactor)
-			if userFactor.FactorType == "push" {
-				factorFound = true
-				break
+			switch userFactor.FactorType {
+			case "push":
+				if passcode != "" {
+					// push does not use passcode
+					continue
+				}
+
+				// FactorType "push" requires 2 steps to verify the factor:
+				// https://developer.okta.com/docs/reference/api/factors/#issue-a-push-factor-challenge
+				// https://developer.okta.com/docs/reference/api/factors/#verify-a-push-factor-challenge
+				return c.validateOktaPush(ctx, oktaClient, orgURL, user, userFactor)
+			case "token:software:totp", "token:hotp", "token", "token:hardware":
+				if passcode == "" {
+					// token* requires passcode
+					continue
+				}
+
+				// FactorTypes that require only one API call with a PassCode param.
+				// https://developer.okta.com/docs/reference/api/factors/#factors-that-require-only-a-verification-operation
+				err = c.validateOktaPassCode(ctx, oktaClient, user, userFactor, passcode)
+				if err != nil {
+					passcodeUserFactorErrors[userFactor] = err
+				} else {
+					// MFA token-type factor succeeded
+					return nil
+				}
 			}
 		}
 	}
 
-	if !factorFound {
+	if attemptedFactors := len(passcodeUserFactorErrors); attemptedFactors > 0 {
+		errs := []error{fmt.Errorf("okta authentication failed after trying %d factors", attemptedFactors)}
+		for userFactor, err = range passcodeUserFactorErrors {
+			errs = append(errs, fmt.Errorf("failed to verify MFA passcode using '%s' factor %s: %w", userFactor.FactorType, userFactor.Id, err))
+		}
+		return errors.Join(errs...)
+	}
+
+	if passcode == "" {
 		return fmt.Errorf("no push-type MFA factor found for user")
 	}
+	return fmt.Errorf("no token-type MFA factors found for user")
+}
 
-	result, _, err := client.UserFactor.VerifyFactor(ctx, user.Id, userFactor.Id, okta.VerifyFactorRequest{}, userFactor, nil)
+func (c *Core) validateOktaPush(ctx context.Context, oktaClient *okta.Client, orgURL *url.URL, user *okta.User, userFactor *okta.UserFactor) error {
```

Comment:
I used separate `validateOktaPush` and `validateOktaPassCode` methods because Push verification has a more complex API flow. At first I started adding `if passcode != ""` conditionals over and over, but decided it would be easier to just make a separate method for token/passcode auth.